src/main/java/com/epam/aidial/cfg/service/transfer/CoreConfigVersionAutoDetectService.java

```diff
@@ -0,0 +1,86 @@
+
+    private static final String CURRENT_VERSION_CACHE_KEY = "current-version";
+
+    private final CoreConfigClient coreConfigClient;
+    private final CoreConfigVersionProperties coreConfigVersionProperties;
+    private final Cache<String, String> versionCache;
+    private final Cache<String, String> nonExpiringVersionCache;
+
+    public CoreConfigVersionAutoDetectService(CoreConfigClient coreConfigClient, CoreConfigVersionProperties coreConfigVersionProperties) {
+        this.coreConfigClient = coreConfigClient;
+        this.coreConfigVersionProperties = coreConfigVersionProperties;
+        this.versionCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(coreConfigVersionProperties.getCacheExpirationMs(), TimeUnit.MILLISECONDS)
+                .build();
+        this.nonExpiringVersionCache = CacheBuilder.newBuilder().build();
+    }
+
+    /**
+     * Gets the Core version with retry mechanism.
+     * If all retries fail, falls back to the target version from properties.
+     * If auto-detection is disabled, returns the target version directly.
+     *
+     * @return the Core version
+     */
+    public String getVersion() {
+        if (!coreConfigVersionProperties.isAutoDetectEnabled()) {
+            String target = coreConfigVersionProperties.getTarget();
+            log.debug("Core version auto-detection is disabled. Using target version: {}", target);
+            return target;
+        }
+
+        String version = versionCache.getIfPresent(CURRENT_VERSION_CACHE_KEY);
+        if (version != null) {
+            log.debug("Using cached Core version: {}", version);
+            return version;
+        }
+
+        try {
+            version = getVersionFromCore();
+            versionCache.put(CURRENT_VERSION_CACHE_KEY, version);
+            nonExpiringVersionCache.put(CURRENT_VERSION_CACHE_KEY, version);
+            return version;
+        } catch (Exception e) {
+            log.warn("Failed to get Core version. Reason: {}", e.getMessage());
+            return fallbackToTargetVersion(e);
+        }
+    }
+
+    private String getVersionFromCore() {
+        log.debug("Attempting to get version from Core");
+        String version = coreConfigClient.getVersion();
+        log.info("Successfully retrieved Core version: {}", version);
+        return version;
+    }
+
+    private String fallbackToTargetVersion(Exception lastException) {
+        String version = nonExpiringVersionCache.getIfPresent(CURRENT_VERSION_CACHE_KEY);
+        if (version != null) {
+            log.debug("Using the last successfully retrieved Core version: {}", version);
+            return version;
+        }
+
+        String targetVersion = coreConfigVersionProperties.getTarget();
+        versionCache.put(CURRENT_VERSION_CACHE_KEY, targetVersion);
+        log.warn("All attempts to get Core version failed, falling back to target version: {}", targetVersion);
+        log.debug("Last exception: ", lastException);
```

Comment:
do you still need `lastException` param in this method?

Reply:
Yes. In 'catch' block we log exception message only, on 'warn' level. Here, we log Throwable with stacktrace, which can be useful, on 'debug' level

Reply:
where do we use such pattern in current project? 
usually we just log.warn full exception at start of catch block.

Reply:
I think that's a good pattern to introduce. Full stacktrace may not be very useful in most cases here since we call client class directly. For example, if service is unavailable or JWT token is invalid (most common cases, I guess), we will see a warn message without messy stacktrace

Reply:
refactored logging & added AnonymousCoreConfigClient to get version (without Auth config)